### PR TITLE
Validate teacher application primary email matches account email

### PR DESCRIPTION
**Pull request by aoby**
[Task](https://trello.com/c/jMomnv39/38-validate-teacher-application-primary-email-matches-account-and-auto-upgrade-to-teacher-account).

This is to solve two problems we've been having:
 1. Teacher signed into someone else's account
 2. Teacher signed up as a student account

Now, for actual teacher accounts, we will initially suggest the primary email from their account settings:
image

If they modify it, we display an error with a link to account settings:
image

Since we only store MD5 hashes of student email addresses, we cannot suggest the email for accidental student accounts:
image

But we can still validate:
image

And, accept when it's a match:
image

In the case of accidental student accounts, submitting the teacher application will automatically (silently) upgrade the account to teacher and set the email.

I also cleaned up some of the teacher application React unit tests while I was in there.

### CODE REVIEW

**Comment by ashercodeorg** [comment]
Is it possible to remove the level changes from the PR?

**Comment by aoby** [comment]
@ashercodeorg that was accidentally swept up since I was based off a very outdated staging. I just rebased.

File: dashboard/app/models/pd/teacher_application.rb [hunk 1]
```diff
@@ -86,6 +86,22 @@ def validate_required_application_fields
...
+      errors.add(
+        :primary_email,
+        "must match your login. If you want to use this email instead, first update it in #{account_settings_link}."
+      )
+    end
+  end
+
+  after_create :ensure_user_is_a_teacher
+  def ensure_user_is_a_teacher
+    user.update!(user_type: User::TYPE_TEACHER, email: primary_email) if user.email.blank?
```
Comment (ashercodeorg) [inline dashboard/app/models/pd/teacher_application.rb:102]: Do you want to throw here if the `User` is invalid after the update? My hope is that we do (I'd like to see the issue surfaced in HB). Realize that (IIRC) we have a bunch of invalid users (beyond missing email).

**Comment by aoby** [comment]
hmm, but that didn't fix it :( I'm not sure where that all came from. Fixing...

**Comment by aoby** [comment]
ok, I removed the extra files and rebased. PTAL

File: dashboard/app/models/pd/teacher_application.rb (hunk 1 above)
Comment (aoby) [reply dashboard/app/models/pd/teacher_application.rb:102]: I agree. I'd rather we be notified of these. Hopefully teachers who are filling out applications have valid accounts (or else they might encounter other issues?)

Comment (aoby) [reply dashboard/app/models/pd/teacher_application.rb:102]: And to be clear, `update!` (with the bang) does throw on invalid.

Comment (ashercodeorg) [reply dashboard/app/models/pd/teacher_application.rb:102]: Right, I wasn't certain if your throwing was deliberate.

File: dashboard/app/models/pd/teacher_application.rb [hunk 2]
```diff
@@ -86,6 +86,22 @@ def validate_required_application_fields
...
+      account_settings_link = ActionController::Base.helpers.link_to('account settings', '/users/edit')
+      errors.add(
+        :primary_email,
+        "must match your login. If you want to use this email instead, first update it in #{account_settings_link}."
+      )
+    end
+  end
+
+  after_create :ensure_user_is_a_teacher
+  def ensure_user_is_a_teacher
```
Comment (ashercodeorg) [inline dashboard/app/models/pd/teacher_application.rb:101]: A test for this would be nice.

**Comment by ashercodeorg** [comment]
LGTM (ruby). Someone else should look at the JS.

File: dashboard/app/models/pd/teacher_application.rb (hunk 2 above)
Comment (aoby) [reply dashboard/app/models/pd/teacher_application.rb:101]: done

File: apps/src/code-studio/pd/teacher_application/teacher_application.jsx [hunk 3]
```diff
@@ -83,23 +87,48 @@ const isUnder1000Chars = (value) => {
...
     schoolDistrictData: React.PropTypes.object.isRequired,
-    districtErrorMessageHandler: React.PropTypes.func.isRequired
+    districtErrorMessageHandler: React.PropTypes.func.isRequired,
+    accountEmail: React.PropTypes.string,
+    hashedAccountEmail: React.PropTypes.string.isRequired
+  },
+
+  componentWillMount() {
+    // Validate primary email is a valid email string, and that it matches the account email.
+    // In case it's an accidental student account (which will be converted to teacher on submit)
```
Comment (mehalshah) [inline apps/src/code-studio/pd/teacher_application/teacher_application.jsx:103]: Explain in more detail why this needs props in the comment

File: apps/src/code-studio/pd/teacher_application/teacher_application.jsx [hunk 4]
```diff
@@ -83,23 +87,48 @@ const isUnder1000Chars = (value) => {
...
+    // compare with the hashed email.
+    const validatePrimaryEmail = (value) => {
+      const isEmailError = isEmail(value);
+      if (isEmailError) {
+        return isEmailError;
+      }
+
+      if (MD5(value).toString() !== this.props.hashedAccountEmail) {
```
Comment (mehalshah) [inline apps/src/code-studio/pd/teacher_application/teacher_application.jsx]: hashes are case sensitive. Convert this to lowercase (like we do [here](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/user.rb#L382))

**Comment by mehalshah** [comment]
🚢  :it:

File: apps/src/code-studio/pd/teacher_application/teacher_application.jsx (hunk 4 above)
Comment (ashercodeorg) [reply apps/src/code-studio/pd/teacher_application/teacher_application.jsx]: Nice catch.

**Pull request merged**